File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/polar/BlePolarDeviceCapabilitiesUtility.kt
package com.polar.androidcommunications.api.ble.model.polar

import android.content.Context
import android.os.Environment
import android.util.Log
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import com.polar.androidcommunications.api.ble.BleLogger
import java.io.File

class BlePolarDeviceCapabilitiesUtility {

    enum class FileSystemType {
        UNKNOWN_FILE_SYSTEM,
        H10_FILE_SYSTEM,
        POLAR_FILE_SYSTEM_V2
    }

    data class DeviceCapabilitiesConfig(
        @SerializedName("version") val version: String = "1.0",
        @SerializedName("devices") val devices: Map<String, DeviceCapabilities> = emptyMap(),
        @SerializedName("defaults") val defaults: DefaultsSection = DefaultsSection()
    )

    data class DeviceCapabilities(
        @SerializedName("fileSystemType") val fileSystemType: String? = null,
        @SerializedName("recordingSupported") val recordingSupported: Boolean? = null,
        @SerializedName("firmwareUpdateSupported") val firmwareUpdateSupported: Boolean? = null,
        @SerializedName("isDeviceSensor") val isDeviceSensor: Boolean? = null,
        @SerializedName("activityDataSupported") val activityDataSupported: Boolean? = null
    )

    data class DefaultsSection(
        @SerializedName("fileSystemType") val fileSystemType: String = "POLAR_FILE_SYSTEM_V2",
        @SerializedName("recordingSupported") val recordingSupported: Boolean = false,
        @SerializedName("firmwareUpdateSupported") val firmwareUpdateSupported: Boolean = true,
        @SerializedName("isDeviceSensor") val isDeviceSensor: Boolean = false,
        @SerializedName("activityDataSupported") val activityDataSupported: Boolean = false
    )

    companion object {

        private const val TAG = "BlePolarDeviceCapabilitiesUtility"
        private const val FILE_NAME = "polar_device_capabilities.json"

        private val gson = Gson()
        private lateinit var config: DeviceCapabilitiesConfig
        private var initialized = false

        /**
         * Initializes the device capabilities configuration.
         *
         * This method first attempts to read the configuration JSON from the public
         * Documents/PolarConfig folder, allowing the config to be overridden at runtime.
         * If that file does not exist, it copies the bundled default from the app's assets.
         *
         * If external storage is unavailable (e.g. permission not granted), the method falls
         * back to reading the bundled asset directly. This guarantees that the SDK is always
         * initialized regardless of whether the user has granted storage permission.
         *
         * The configuration is parsed and stored in memory for subsequent API calls.
         * This method is safe to call multiple times and will only initialize once.
         *
         * @param context application or activity context
         */
        @JvmStatic
        @Synchronized
        fun initialize(context: Context) {
            if (initialized) return

            // Try loading from the user-overridable external config first

            try {

                val assetJson = context.assets.open(FILE_NAME)
                    .bufferedReader()
                    .use { it.readText() }

                val assetConfig =
                    gson.fromJson(assetJson, DeviceCapabilitiesConfig::class.java)

                val folder = File(
                    Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS),
                    "PolarConfig"
                )
                if (folder.exists() || folder.mkdirs()) {
                    val configFile = File(folder, FILE_NAME)

                    if (!configFile.exists()) {
                        configFile.writeText(assetJson)
                        Log.d(TAG, "Default config copied to $configFile")
                    }

                    val userJson = configFile.inputStream().bufferedReader().use { it.readText() }
                    val userConfig = gson.fromJson(userJson, DeviceCapabilitiesConfig::class.java)

                    if (assetConfig.version != userConfig.version) {
                        BleLogger.d(TAG, "Config version changed ${userConfig.version} → ${assetConfig.version}, migrating")

                        val merged = mergeConfigs(userConfig, assetConfig)

                        configFile.writeText(gson.toJson(merged))
                        config = merged
                    } else {
                        config = userConfig
                    }

                    initialized = true
                    Log.d(TAG, "BlePolarDeviceCapabilitiesUtility initialized successfully")
                    return
                }
            } catch (e: Exception) {
                BleLogger.w(
                    TAG,
                    "Could not load capabilities from external storage, falling back to bundled asset: ${e.message}"
                )
            }

            // Fallback: read directly from the bundled AAR asset (no permissions required)
            try {
                val json = context.assets.open(FILE_NAME).bufferedReader().use { it.readText() }
                config = gson.fromJson(json, DeviceCapabilitiesConfig::class.java)
                initialized = true
                Log.d(TAG, "BlePolarDeviceCapabilitiesUtility initialized from bundled asset")
            } catch (e: Exception) {
                BleLogger.e(TAG, "Failed to initialize capabilities from bundled asset: ${e.message}")
            }
        }

        private fun mergeConfigs(
            user: DeviceCapabilitiesConfig,
            asset: DeviceCapabilitiesConfig
        ): DeviceCapabilitiesConfig {

            val mergedDevices = asset.devices.toMutableMap()

            user.devices.forEach { (key, userDevice) ->
                val assetDevice = asset.devices[key]

                mergedDevices[key] = DeviceCapabilities(
                    fileSystemType = userDevice.fileSystemType ?: assetDevice?.fileSystemType,
                    recordingSupported = userDevice.recordingSupported ?: assetDevice?.recordingSupported,
                    firmwareUpdateSupported = userDevice.firmwareUpdateSupported ?: assetDevice?.firmwareUpdateSupported,
                    isDeviceSensor = userDevice.isDeviceSensor ?: assetDevice?.isDeviceSensor,
                    activityDataSupported = userDevice.activityDataSupported ?: assetDevice?.activityDataSupported
                )
            }

            return DeviceCapabilitiesConfig(
                version = asset.version,
                devices = mergedDevices,
                defaults = asset.defaults
            )
        }

        private fun ensureInitialized(): Boolean {
            if (!initialized) {
                BleLogger.w(
                    TAG,
                    "BlePolarDeviceCapabilitiesUtility used before initialize()"
                )
                return false
            }
            return true
        }

        /**
         * Get type of filesystem the device supports.
         *
         * @param deviceType device type
         * @return type of the file system supported or unknown file system type
         */
        @JvmStatic
        fun getFileSystemType(deviceType: String): FileSystemType {
            if (!ensureInitialized()) return FileSystemType.UNKNOWN_FILE_SYSTEM

            val fs = config.devices[deviceType.lowercase()]?.fileSystemType
                ?: config.defaults.fileSystemType

            val result = when (fs) {
                "H10_FILE_SYSTEM" -> FileSystemType.H10_FILE_SYSTEM
                "POLAR_FILE_SYSTEM_V2" -> FileSystemType.POLAR_FILE_SYSTEM_V2
                else -> FileSystemType.UNKNOWN_FILE_SYSTEM
            }

            BleLogger.d(TAG, "getFileSystemType($deviceType) -> $result")
            return result
        }

        /**
         * Check if device supports recording start and stop.
         *
         * @param deviceType device type
         * @return true if device supports recording
         */
        @JvmStatic
        fun isRecordingSupported(deviceType: String): Boolean {
            if (!ensureInitialized()) return false

            val result = config.devices[deviceType.lowercase()]?.recordingSupported
                ?: config.defaults.recordingSupported
            BleLogger.d(TAG, "isRecordingSupported($deviceType) -> $result")
            return result
        }

        /**
         * Check if device supports firmware update.
         *
         * @param deviceType device type
         * @return true if firmware update is supported
         */
        @JvmStatic
        fun isFirmwareUpdateSupported(deviceType: String): Boolean {
            if (!ensureInitialized()) return false

            val result = config.devices[deviceType.lowercase()]?.firmwareUpdateSupported
                ?: config.defaults.firmwareUpdateSupported

            BleLogger.d(TAG, "isFirmwareUpdateSupported($deviceType) -> $result")
            return result
        }

        /**
         * Check if device is considered a sensor device.
         *
         * @param deviceType device type
         * @return true if device is a sensor
         */
        @JvmStatic
        fun isDeviceSensor(deviceType: String): Boolean {
            if (!ensureInitialized()) return false

            val result = config.devices[deviceType.lowercase()]?.isDeviceSensor
                ?: config.defaults.isDeviceSensor

            BleLogger.d(TAG, "isDeviceSensor($deviceType) -> $result")
            return result
        }

        /**
         * Check if device supports activity data storage and sync.
         *
         * @param deviceType device type
         * @return true if activity data is supported
         */
        @JvmStatic
        fun isActivityDataSupported(deviceType: String): Boolean {
            if (!ensureInitialized()) return false

            val result = config.devices[deviceType.lowercase()]?.activityDataSupported
                ?: config.defaults.activityDataSupported

            BleLogger.d(TAG, "isActivityDataSupported($deviceType) -> $result")
            return result
        }
    }
}

File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/polar/BlePolarDeviceCapabilitiesUtility.swift
import Foundation

open class BlePolarDeviceCapabilitiesUtility {
    public enum FileSystemType {
        case unknownFileSystem
        case h10FileSystem
        case polarFileSystemV2
    }

    private static let fileName = "polar_device_capabilities.json"
    private static let configSubfolder = "PolarConfig"
    private static let lock = NSLock()
    private static var initialized = false
    private static var capabilities: [String: DeviceCapabilities] = [:]
    private static var defaults: DeviceCapabilities?

    // MARK: - Codable structs

    private struct DeviceCapabilities: Codable {
        let fileSystemType: String?
        let recordingSupported: Bool?
        let firmwareUpdateSupported: Bool?
        let activityDataSupported: Bool?
        let isDeviceSensor: Bool?
    }

    private struct DeviceCapabilitiesConfig: Codable {
        let version: String?
        let devices: [String: DeviceCapabilities]
        let defaults: DeviceCapabilities
    }

    // MARK: - Bundle access

    private static func loadBundledData() -> Data? {
        // Priority 1: app's main bundle.
        // An app can ship its own polar_device_capabilities.json in its target resources
        // to override the SDK's baseline entirely. This works consistently whether the SDK
        // is integrated via CocoaPods or SPM.
        if let url = Bundle.main.url(forResource: "polar_device_capabilities", withExtension: "json"),
           let data = try? Data(contentsOf: url) {
            return data
        }
        // Priority 2: SDK's own bundled baseline.
        // SPM synthesises Bundle.module pointing at the package resource bundle.
        // CocoaPods use_frameworks! puts resources in the class's framework bundle.
        // SWIFT_PACKAGE is set automatically by the SPM build system.
#if SWIFT_PACKAGE
        if let url = Bundle.module.url(forResource: "polar_device_capabilities", withExtension: "json"),
           let data = try? Data(contentsOf: url) {
            return data
        }
#else
        if let url = Bundle(for: BlePolarDeviceCapabilitiesUtility.self).url(forResource: "polar_device_capabilities", withExtension: "json"),
           let data = try? Data(contentsOf: url) {
            return data
        }
#endif
        return nil
    }

    // MARK: - Initialisation

    /// Initialises the device capabilities configuration.
    ///
    /// 1. Loads the bundled JSON as the baseline configuration.
    /// 2. Attempts to use `Documents/PolarConfig/polar_device_capabilities.json` in the app sandbox:
    ///    - Seeds the file from the bundle if it does not yet exist.
    ///    - If the on-disk `version` differs from the bundled one, user-supplied entries are merged
    ///      into the bundled config and the sandbox file is updated.
    ///    - If the versions match the sandbox file is used as-is.
    /// 3. Falls back to the bundled JSON if any filesystem step fails.
    ///
    /// Safe to call multiple times; initialises only once.
    public static func initialize() {
        lock.lock()
        defer { lock.unlock() }

        guard !initialized else { return }

        guard let bundledData = loadBundledData() else {
            BleLogger.trace("BlePolarDeviceCapabilitiesUtility: Bundled polar_device_capabilities.json not found — capabilities unavailable")
            return
        }

        let decoder = JSONDecoder()
        guard let bundledConfig = try? decoder.decode(DeviceCapabilitiesConfig.self, from: bundledData) else {
            BleLogger.trace("BlePolarDeviceCapabilitiesUtility: Failed to decode bundled capabilities JSON")
            return
        }

        do {
            let fileManager = FileManager.default
            let documentsURL = try fileManager.url(
                for: .documentDirectory,
                in: .userDomainMask,
                appropriateFor: nil,
                create: true
            )
            let configFolder = documentsURL.appendingPathComponent(configSubfolder)
            let configURL = configFolder.appendingPathComponent(fileName)

            if !fileManager.fileExists(atPath: configFolder.path) {
                try fileManager.createDirectory(at: configFolder, withIntermediateDirectories: true)
            }

            if !fileManager.fileExists(atPath: configURL.path) {
                try bundledData.write(to: configURL)
                BleLogger.trace("BlePolarDeviceCapabilitiesUtility: Seeded sandbox config at \(configURL.path)")
                applyConfig(bundledConfig)
                return
            }

            let userData = try Data(contentsOf: configURL)
            guard let userConfig = try? decoder.decode(DeviceCapabilitiesConfig.self, from: userData) else {
                BleLogger.trace("BlePolarDeviceCapabilitiesUtility: Failed to decode sandbox capabilities JSON — using bundled version")
                applyConfig(bundledConfig)
                return
            }

            if userConfig.version != bundledConfig.version {
                BleLogger.trace("BlePolarDeviceCapabilitiesUtility: Version changed \(userConfig.version ?? "nil") → \(bundledConfig.version ?? "nil"), merging")
                let merged = mergeConfigs(user: userConfig, bundled: bundledConfig)
                let encoder = JSONEncoder()
                encoder.outputFormatting = .prettyPrinted
                if let mergedData = try? encoder.encode(merged) {
                    try? mergedData.write(to: configURL)
                }
                applyConfig(merged)
            } else {
                applyConfig(userConfig)
            }

            BleLogger.trace("BlePolarDeviceCapabilitiesUtility: Initialized successfully from \(configURL.path)")
        } catch {
            BleLogger.trace("BlePolarDeviceCapabilitiesUtility: Filesystem error (\(error)) — falling back to bundled config")
            applyConfig(bundledConfig)
        }
    }

    private static func applyConfig(_ config: DeviceCapabilitiesConfig) {
        defaults = config.defaults
        capabilities = config.devices
        initialized = true
    }

    // MARK: - Merge

    /// Merges user-supplied device entries on top of the bundled baseline.
    /// User-specified non-nil values win; new devices added in the bundle are carried over.
    private static func mergeConfigs(
        user: DeviceCapabilitiesConfig,
        bundled: DeviceCapabilitiesConfig
    ) -> DeviceCapabilitiesConfig {
        var mergedDevices = bundled.devices
        for (key, userDevice) in user.devices {
            let bundledDevice = bundled.devices[key]
            mergedDevices[key] = DeviceCapabilities(
                fileSystemType: userDevice.fileSystemType ?? bundledDevice?.fileSystemType,
                recordingSupported: userDevice.recordingSupported ?? bundledDevice?.recordingSupported,
                firmwareUpdateSupported: userDevice.firmwareUpdateSupported ?? bundledDevice?.firmwareUpdateSupported,
                activityDataSupported: userDevice.activityDataSupported ?? bundledDevice?.activityDataSupported,
                isDeviceSensor: userDevice.isDeviceSensor ?? bundledDevice?.isDeviceSensor
            )
        }
        return DeviceCapabilitiesConfig(
            version: bundled.version,
            devices: mergedDevices,
            defaults: bundled.defaults
        )
    }

    // MARK: - Lazy-init helper

    private static func ensureInitialized() -> Bool {
        if !initialized { initialize() }
        return initialized
    }

    // MARK: - Public API

    /// Get type of filesystem the device supports.
    /// - Parameter deviceType: device type (case-insensitive)
    /// - Returns: filesystem type, or `.unknownFileSystem` if unavailable
    public static func fileSystemType(_ deviceType: String) -> FileSystemType {
        guard ensureInitialized() else { return .unknownFileSystem }
        let fsType = (capabilities[deviceType.lowercased()]?.fileSystemType ?? defaults?.fileSystemType)?.uppercased()
        switch fsType {
        case "H10_FILE_SYSTEM":    return .h10FileSystem
        case "POLAR_FILE_SYSTEM_V2": return .polarFileSystemV2
        default:                   return .unknownFileSystem
        }
    }

    /// Check if device supports recording start and stop over BLE.
    /// - Parameter deviceType: device type (case-insensitive)
    /// - Returns: true if recording is supported
    public static func isRecordingSupported(_ deviceType: String) -> Bool {
        guard ensureInitialized() else { return false }
        return capabilities[deviceType.lowercased()]?.recordingSupported ?? defaults?.recordingSupported ?? false
    }

    /// Check if device supports firmware update.
    /// - Parameter deviceType: device type (case-insensitive)
    /// - Returns: true if firmware update is supported
    public static func isFirmwareUpdateSupported(_ deviceType: String) -> Bool {
        guard ensureInitialized() else { return false }
        return capabilities[deviceType.lowercased()]?.firmwareUpdateSupported ?? defaults?.firmwareUpdateSupported ?? false
    }

    /// Check if device supports activity data storage and sync.
    /// - Parameter deviceType: device type (case-insensitive)
    /// - Returns: true if activity data is supported
    public static func isActivityDataSupported(_ deviceType: String) -> Bool {
        guard ensureInitialized() else { return false }
        return capabilities[deviceType.lowercased()]?.activityDataSupported ?? defaults?.activityDataSupported ?? false
    }

    /// Check if device is a sensor (as opposed to a watch or training computer).
    /// - Parameter deviceType: device type (case-insensitive)
    /// - Returns: true if device is a sensor
    public static func isDeviceSensor(_ deviceType: String) -> Bool {
        guard ensureInitialized() else { return false }
        return capabilities[deviceType.lowercased()]?.isDeviceSensor ?? defaults?.isDeviceSensor ?? false
    }
}
